pages/_middleware.ts

```diff
@@ -1,23 +1,51 @@
 import { NextRequest, NextResponse } from 'next/server'
 
-const BLOCKED_COUNTRIES = ['US', 'BY', 'CU', 'IR', 'IQ', 'CI', 'LR', 'KP', 'SD', 'SY', 'ZW', 'CA']
-const ALLOWED_URLS = RegExp(
-  '(/images|/favicon.ico|/font|/api/auction/getLatestAuction|/api/auction/getAuctionById|/api/auction/getLastHedge)',
-)
-
-export function middleware(req: NextRequest) {
-  const country = req?.geo?.country
-  const response = NextResponse.next()
-  const path = req.url
-
-  if (country && BLOCKED_COUNTRIES.includes(country) && !ALLOWED_URLS.test(path)) {
-    return NextResponse.rewrite('/blocked')
+import { redis } from '../utils/redisClient'
+import { isVPN } from '../utils/vpn'
+
+const ignoredPaths = ['/api', '/favicon.ico', '/static', '/_next', '/blocked']
+
+export async function middleware(request: NextRequest) {
+  const url = request.nextUrl
+
+  // should not block api calls, static files, nextjs files, favicon, blocked page and files with extension
+  const isIgnoredPath = ignoredPaths.some(path => url.pathname.startsWith(path)) || url.pathname.includes('.')
+  if (isIgnoredPath) {
+    return NextResponse.next()
+  }
+
+  const cloudflareCountry = request.headers.get('cf-ipcountry')
+  const country = cloudflareCountry ?? request?.geo?.country
+
+  const ip = request.headers.get('cf-connecting-ip') || request.headers.get('x-forwarded-for') || request.ip
+
+  const allowedIPs = (process.env.WHITELISTED_IPS || '').split(',')
+  const isIPWhitelisted = ip && allowedIPs.includes(ip)
+
+  if (ip && !isIPWhitelisted) {
+    const redisData = await redis.get(ip)
+    const isIPBlocked = !!redisData
+
+    console.log('ip', ip, isIPBlocked, url.protocol, url.host, '/blocked')
+
+    if (isIPBlocked && url.pathname !== '/blocked') {
+      return NextResponse.redirect(`${url.protocol}//${url.host}/blocked`)
+    }
+
+    const isFromVpn = await isVPN(ip)
```

Comment:
should implement the same things from data center PRs, but can be in separate PR if prefer

Reply:
this commit takes care of it: https://github.com/KMKoushik/squeeth-portal/pull/118/commits/b24ecec64bc61571acd033e09ccaa54444f6f137